Combined patch against OpenSSL_1_1_0-stable branch for the more or less obvious issues

bernd-edlinger opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] CLA is signed

##### Description of change
<!-- Provide a description of the changes.

If it fixes a github issue, add Fixes #XXXX.
-->
Fixes #2103 #2104 #2105 #2109 #2111 #2115
This pull request is for trunk and 1.1.0 branch.

The current fix on trunk for x_name.c is still buggy, i.e
will crash if the first goto err is taken, and has obviously
not got much test coverage, so I would prefer to use my
solution here.

Note: the ex_data.c has a race condition on trunk that
was not there in 1.0.2:  f->free_func can change concurrently
with the call of the function pointer, but that is hard to fix
because it is designed into the new API function CRYPTO_free_ex_index.

I have not found out if the issue #2112 (missing EC_GROUP_free)
is still open or not, I basically was unable to locate the patch context.

All other issues were already fixed on trunk, but not yet back ported.

levitte commented on crypto/ex_data.c:
```diff
@@ -325,20 +326,23 @@ void CRYPTO_free_ex_data(int class_index, void *obj, CRYPTO_EX_DATA *ad)
-    }
     for (i = 0; i < mx; i++) {
-        if (storage[i] && storage[i]->free_func) {
+        if (storage)
```
`if (storage != NULL)`

levitte commented on crypto/ex_data.c:
```diff
@@ -325,20 +326,23 @@ void CRYPTO_free_ex_data(int class_index, void *obj, CRYPTO_EX_DATA *ad)
+            f = sk_EX_CALLBACK_value(ip->meth, i);
+            CRYPTO_THREAD_unlock(ex_data_lock);
+        }
+        if (f && f->free_func) {
```
`if (f  != NULL && f->free_func  != NULL) {`

levitte commented:
Please rebase on top of fresher 1.1.0

levitte commented:
Is this for both master and OpenSSL_1_1_0-stable (1.1.0)?

bernd-edlinger commented:
I rebased it to fresh 1_1_0-stable. And updated the change log.
Initially it did work for master but now I see a conflict in ssl/statem/statem_srvr.c
So I will have to create a new pull request for master.

richsalz commented:
ping @levitte

levitte commented:
Merged

5d0634a93bdc18e753d641c4563cb18f685e148f
